internal/staging/stdscript/addressv0.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package stdscript
+
+import (
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/txscript/v4/stdaddr"
+)
+
+// addrToSlice is a convenience function that returns a slice containing the
+// passed address if the given error is nil and the address is NOT nil.
+func addrToSlice(addr stdaddr.Address, err error) []stdaddr.Address {
+	if err != nil || addr == nil {
+		return nil
+	}
+	return []stdaddr.Address{addr}
+}
+
+// ExtractAddrsV0 analyzes the passed version 0 public key script and returns
+// the associated script type along with any addresses associated with it when
+// possible.
+//
+// Note that it only works for standard script types and any data such as public
+// keys which are invalid are omitted from the results.
+func ExtractAddrsV0(pkScript []byte, params stdaddr.AddressParamsV0) (ScriptType, []stdaddr.Address) {
+	// Check for pay-to-pubkey-hash-ecdsa-secp256k1 script.
+	if h := ExtractPubKeyHashV0(pkScript); h != nil {
+		addr, err := stdaddr.NewAddressPubKeyHashEcdsaSecp256k1V0(h, params)
+		return STPubKeyHashEcdsaSecp256k1, addrToSlice(addr, err)
```

Comment:
General issue on this and the similar tests below. 

It's true that `NewAddressPubKeyHashEcdsaSecp256k1V0` will never (currently) return an error here, since the only error it _could_ return is if `h` has the wrong size, which is guarded by the fact that `ExtractPubKeyHashV0` only returns non-nil if the payload data (hash in this particular case) has the correct size. In fact, capturing the error and calling `addrToSlice` is redundant, since the error _can't_ be nil. 

So this is currently all correct and I don't _expect_ it to change in the future. 

My only small nitpick here is that if any of these assumptions changed in the future, these cases could end up returning a `nil` address slice but a defined `ScriptType` (e.g. `STPubKeyHashEcdsaSecp256k1` in this case with a nil slice instead of `STNonStandard`). 

Then again, re-reading the function's comment, it's not obvious to me now whether it should be the case that this function should return `STNonStandard` if it can't create the address after extracting the payload data. This is easy to cause in the pubkey variations (by encoding a correctly sized but otherwise invalid pubkey).

A simple modification to enforce that would be to add the `ScriptType`  as an argument to `addrToSlice()` and then redefine it to return `(ScriptType, []stdaddr.Address, error)` so that all clauses would consistently return `STNonStandard` in case the `stdaddr` in case the assumptions are broken in future commits.

Reply:
The described behavior is intentional so that the script type matches the results of `DetermineScriptType` even if there is a failure when creating the addresses for some reason.  The existing `ExtractPkScriptAddrs` works this way too and, as you noted, the comment calls out that any invalid data is omitted from the results.  I guess maybe it isn't super clear why that is the case though.

I also agree that the error check is actually redundant currently since it can't fail, but, as you mentioned, assumptions could change in the future where maybe it's possible for a non-nil error to be returned after passing extract, so it is practicing a bit of defensive coding by being paranoid there.

I should also note that there are various places in the code that invoke the existing `ExtractPkScriptAddrs` and, after the conversion, this new version, that depend on this behavior.

Reply:
All of that said, maybe we should consider giving it the same treatment as `DetermineRequiredSigs` and just remove the script type return from it altogether.

Personally, I'd rather keep it for the sake of efficiency since it already necessarily has to determine the script type to create the addresses anyway and, surveying the code, I see that the only consumers that ignore the script type are two cases in tests.  In other words, all of the consumers use the returned script type to avoid another call to `DetermineScriptType`.

However, unlike many years ago, `DetermineScriptType` is now extremely efficient and is all zero alloc nowadays, so the extra call really probably wouldn't even be noticeable in practice.

I'd like to hear if @jrick and @chappjc have any thoughts here since dcrwallet and dcrdata are both primary consumers of this logic.

Reply:
So if you need all the outputs of `ExtractPkScriptAddrs`, you're going to be replacing it with:

- `DetermineScriptType`
- `ExtractAddrs`
- `DetermineRequiredSigs`

If `ExtractAddrs` returns the script type too, we don't need `DetermineScriptType`.  Is that what's being discussed?
It's a strong argument for efficiency IMO to have `ExtractAddrs` return the script class like this.

Reply:
> So if you need all the outputs of `ExtractPkScriptAddrs`, you're going to be replacing it with:

With this PR as it stands, to replace all the outputs of the current `ExtractPkScriptAddrs` you only need `ExtractAddrs` and `DetermineRequiredSigs`.  If we were to remove the script type return from this PR, as this is potentially discussing, then, yes, you'd also need to independently call `DetermineScriptType`.

> If `ExtractAddrs` returns the script type too, we don't need `DetermineScriptType`. Is that what's being discussed?

No.  @matheusd is pointing out that if there is a problem when creating an address after identifying the script type for some reason, this code will still return the script type that was identified despite the error in address creation.  He was saying that he thinks it should return `STNonStandard` in that case.  However, as I called out in my original reply, the implemented behavior is actually intentional so that whatever script type this returns will match the one returned by `DetermineScriptType`.

So, the question really comes down to one of three choices:

1. Keep the behavior as implemented so that it returns a script type that always matches whatever `DetermineScriptType` returns even if there is an issue actually creating addresses and those addresses are just filted.   Note this is also the behavior of the existing `ExtractPkScriptAddrs`.
2. Change the behavior to return `STNonStandard` if there is an issue when creating the new addresses which would mean the script type would no longer necessarily always match what `DetermineScriptType` returns.
3. Remove the script type return altogether to avoid the potential ambiguity by only having `DetermineScriptType` be the single source of truth, so to speak.

The first option is my preference.

I don't like the second option personally because I think have disparate script types would be totally surprising behavior and would almost positively lead to unexpected behavior down the line.

The third option isn't entirely unreasonable, but that's what leads to:

>  It's a strong argument for efficiency IMO to have `ExtractAddrs` return the script class like this.

Agreed and personally I think it should remain, but since it was a concern of @matheusd, I wanted to get feedback from the primary consumers of the API.

Reply:
Maybe just document that the address slice should not be assumed to be non-zero (or any particular length) for a recognized script type, and the caller should always check the length in case the addresses were not able to be decoded.

More error returns at not what I want, so maybe if there is ever a conceivable error that the caller might want to log or otherwise inspect, that can be added in the future?

Reply:
> Maybe just document that the address slice should not be assumed to be non-zero (or any particular length) for a recognized script type, and the caller should always check the length in case the addresses were not able to be decoded.

It currently says:
```
// Note that it only works for standard script types and any data such as public
// keys that are invalid are omitted from the results.
//
```

I can add the additional suggested blurb.

> If the caller must know the issue, they at least get the script type so that they can call the appropriate NewAddress function to identify the failure.

That is my take as well.  I did a survey of all current callers when implementing this and everything is already properly handling it because it's the existing behavior too.

Reply:
Updated to call out:

```
// This function only works for standard script types and any data that fails to
// produce a valid address is omitted from the results.  This means callers must
// not blindly assume the slice will be of a particular length for a given
// returned script type and should always check the length prior to access in
// case the addresses were not able to be decoded.
```